### PR TITLE
util: refactor mechanism used for atomic creation of http code stat-names to a general utility

### PR DESCRIPTION
**Pull request by jmarantz**
Commit Message: Adds a utility class to facilitate making lazy-initialized shared structures across threads, where the lazy-initialization doesn't require a lock except when creation is required.
Additional Description: 
Risk Level: low -- this is a refactor of an existing mechanism.
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by jmarantz** [comment]
This is a prep-step toward the resolution we discussed for https://github.com/envoyproxy/envoy/issues/11069

**Comment by jmarantz** [comment]
This coverage failure appears to be caused by this PR but I don't understand why yet. asan, tsan, and valgrind all think the newly added test is fine. Suspect g++ library issue? Failures seem to be in tcmalloc.

**Comment by jmarantz** [comment]
Note to self: when there's a coverage failure, check for test-class name clashes. That was the case this time.

File: source/common/common/thread.h [hunk 1]
```diff
@@ -80,24 +83,28 @@ enum class AtomicPtrAllocMode { DoNotDelete, DeleteOnDestruct };
...
     }
   }
 
   ~AtomicPtrArray() {
     if (alloc_mode == AtomicPtrAllocMode::DeleteOnDestruct) {
-      for (auto& ptr : data_) {
-        delete ptr;
+      for (std::atomic<T*>& atomic_ref : data_) {
+        T* ptr = atomic_ref.load();
+        if (ptr != nullptr) {
```
Comment (mattklein123) [inline source/common/common/thread.h:105]: This is not actually thread safe. Can you add more comments as to why this is OK based on usage or some other factor? Or should the mutex be held. Also, should data_ have lock annotations on it?

Comment (jmarantz) [reply source/common/common/thread.h:105]: By the time we are destructing the object, there better be only one thread accessing it or mutexes cannot help :)

https://stackoverflow.com/questions/646427/should-destructors-be-threadsafe

Comment (mattklein123) [reply source/common/common/thread.h:105]: OK sorry I had it in my head that we are taking a reference to a vector and using it, but this makes sense.

**Pull request merged**